### PR TITLE
Add ObjectStoreScheme (#4047)

### LINKED ISSUE #4047: object_store: Instantiate object store from provided url with store options
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently, in the projects that are using `object_store` - datafusion/ delta-rs/ pola-rs etc, a `dyn ObjectStore` has to be created manually by parsing the provided URL, checking the scheme and providing the options.

It would be great to have this capability directly provided by the crate.

**Describe the solution you'd like**
My proposal is to standardize this implementation and bring it into this crate itself exposed by a simple function like the below as it would make things significantly simple for developers using the crate.

```
/// Taken a reference from implementation in delta-rs 
https://github.com/delta-io/delta-rs/blob/c8371b38fdf22802f0f91b4ddc2a47da6be97c68/rust/src/storage/mod.rs#L85-L100

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct StorageOptions(pub HashMap<String, String>);

#[derive(Debug, Clone)]
pub struct AObjectStore {
    storage: Arc<dyn ObjectStore>,
    location: Url,
    options: StorageOptions,
}

/// Try creating a new instance of [`AObjectStore`]
pub fn get_object_store(location: Url, options: impl Into<StorageOptions> + Clone) -> Result<AObjectStore> {
    let prefix = Path::from(location.path());
   
    // parse URL to a kind (s3/ aws/ ... )
    let kind = ObjectStoreKind::parse_url(&location)?;

    // instantiate object store
    let store = kind.into_impl( .... );

    // return
    Ok(Self {
        store,
        location,
        options: options.into(),
    })
}
```

For any new storage backends that may come up in the future, they can be added to the `ObjectStoreKind` along with a small implementation in `into_impl`. Users of the crate will only have to bump up the crate version. 

**Describe alternatives you've considered**
Without this, each lib using `object_store` has to implement its own parsing.

Examples: 
- See datafusion registry [here](https://github.com/apache/arrow-datafusion/blob/52fa2285b43ad6712e9b8bf6c05b4b8ff93f44f9/datafusion/execution/src/object_store.rs#L186-L217).
- See delta-rs implementation [here](https://github.com/delta-io/delta-rs/blob/c8371b38fdf22802f0f91b4ddc2a47da6be97c68/rust/src/storage/config.rs#L138-L196)

Also without this each time this crate adds a new backend, the users of this crate will have to bump up the version and add implementation for the backends by themselves. 

**Additional context**
This idea is also implemented by,

- PyArrow FileSystem API [fs.FileSystem.from_uri(uri)](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.from_uri)
- Hadoop FileSystem API [org.apache.hadoop.fs.FileSystem.get(uri, conf)](https://hadoop.apache.org/docs/r3.0.0/api/org/apache/hadoop/fs/FileSystem.html#get-java.net.URI-org.apache.hadoop.conf.Configuration-)
- Fsspec [fsspec.filesystem(protocol, **storage_options)](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.filesystem)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4047

# Rationale for this change
 

Inspired by the work in #4077 I started work on a generic `ObjectStoreBuilder`, however, this ran into a couple of challenges

* Different stores have different configuration APIs, e.g. S3 has ClientOptions, whereas LocalFilesystem has a prefix
* Some downstreams will want to use environment variables, and some will not
* Some downstreams will want to filter config to just match those expected by the store, others will want to error
* There are various valid ways LocalFileSystem and Memory URLs could be interpreted
* The feature flag plumbing was extremely fiddly, the fact we separate the different stores is a historical artifact from when they brought in different SDKs, downstreams shouldn't need to do this

Whilst this PR does to a certain extent punt the issue onto downstreams, I think this standardises the non-trivial URL recognition logic, whilst allowing downstreams to make their own opinionated decision w.r.t the above.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/scheme.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+use crate::{Error, Result};
+use url::Url;
+
+/// Recognises various URL formats, identifying the relevant [`ObjectStore`](crate::ObjectStore)
+///
+/// This can be combined with the [with_url](crate::aws::AmazonS3Builder::with_url) methods
+/// on the corresponding builder to construct the relevant type of store
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ObjectStoreScheme {
```
Comment (alamb) [inline object_store/src/scheme.rs:26]: I don't fully understand why you are proposing to add this to the object store crate.

Users of `object_store` would still have to match on the resulting scheme and instantiate a builder / configuration appropriate to whatever they wanted. The extra value to having a hard coded list of url prefixes seems relatively minimal.

Maybe this is just a first step. 

If I were a user I would want something that took a url like `s3://foo-bucket` or `https://andrew:lamb@foo.com/path` and returned an `Arc<dyn ObjectStore>` . 

For convenience the object_store crate could have default interpretations of these urls, but also some way to extend the API;

Basically I think the API here makes a lot of sense https://docs.rs/datafusion/latest/datafusion/datasource/object_store/trait.ObjectStoreRegistry.html

Comment (alamb) [reply object_store/src/scheme.rs:26]: > This API is just a trait? It doesn't contain any parsing logic, nor any logic to interpret schemes directly

The API is a trait but DataFusion provides default parsing logic / scheme interpretation in https://docs.rs/datafusion/latest/datafusion/datasource/object_store/struct.DefaultObjectStoreRegistry.html


> As explained in the description, because there isn't a way I can see to make such an API that is both coherent and not extremely opinionated...

I wast trying to suggest an API that let users implement their own opinions while also providing a default implementation that worked for simple cases (with whatever opinions you wanted)

**Comment by tustvold** [comment]
I will add a build_with_options method to this
